### ISSUE TITLE
The sanitizer is removing HTML when the tag depth is more than 256 sanitizer removes the tags which actually has the content

### ISSUE THREAD

**Issue by rasmitam** [OWASP/java-html-sanitizer#205]
`I` have a HTML which is surrounded by a big list of empty `<div>` tags, sanitizer is removing a portion of a HTML. When removed the empty list of `<div>` tags from the HTML the sanitizer did not strip off the portion of the HTML.
Please find the HTML where the issue is seen. Please suggest what can be done for this issue.

```
<div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div dir="rtl"><div class="gmail_quote"><div class="gmail_attr" style="text-align:right" dir="ltr"><b><span lang="HE" style="font-family:&quot;David&quot;,&quot;sans-serif&quot;;font-size:14pt">,חבר יקר</span></b></div><div dir="rtl"><div dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div lang="EN-US">Rasmita01<div><div><div><div>Rasmita02<div><div><div><div><div>Rasmita03<div>Rasmita04<div><div><div><div>Rasmita05<div><div><div><div>Rasmita06<div>Rasmita07<div>Rasmita08<div>Rasmita09<div>Rasmita010<div><div><div><div><div><div>Rasmita011<div><div>Rasmita012<div><div><div>Rasmita00<div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div>Rasmita000<div>Rasmita0000<div><div><div>Rasmita1 \r\n </div><div></div>Testing Sanitizer is removing divs when the number of open tags are very high<div></div><div><p class="MsoNormal" style="text-align:right;unicode-bidi:embed;direction:rtl" dir="RTL"><span lang="HE"> <u></u><u></u></span></p></div><div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div lang="EN-US"></div><div lang="EN-US"></div><div lang="EN-US"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote"></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div> 
```

Post sanitization

```
<div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div class="gmail_quote"><div dir="rtl"><div dir="rtl"><div class="gmail_quote"><div class="gmail_attr" style="text-align:right" dir="ltr"><b><span lang="HE" style="font-family:&#39;david&#39; , &#39;sans-serif&#39;;font-size:14pt">,חבר יקר</span></b></div><div dir="rtl"><div dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div class="gmail_quote" dir="rtl"><div lang="EN-US">Rasmita01<div><div><div><div>Rasmita02<div><div><div><div><div>Rasmita03<div>Rasmita04<div><div><div><div>Rasmita05<div><div><div><div>Rasmita06<div><div><div><div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div lang="EN-US"></div><div lang="EN-US"></div><div lang="EN-US"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote" dir="rtl"></div></div><div class="gmail_quote"></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div></div> 
```

Post sanitization:
After Rasmita06 all other divs are removed.

What I did next is removed the unwanted divs from the top and also removed the closing divs and see the results below.

```
<div lang="EN-US">Rasmita01<div><div><div><div>Rasmita02<div><div><div><div><div>Rasmita03<div>Rasmita04<div><div><div><div>Rasmita05<div><div><div><div>Rasmita06<div>Rasmita07<div>Rasmita08<div>Rasmita09<div>Rasmita010<div><div><div><div><div><div>Rasmita011<div><div>Rasmita012<div><div><div>Rasmita00<div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div>Rasmita000<div>Rasmita0000<div><div><div>Rasmita1 \r\n </div><div></div>Testing Sanitizer is removing divs when the number of open tags are very high<div></div><div><p class="MsoNormal" style="text-align:right;unicode-bidi:embed;direction:rtl" dir="RTL"><span lang="HE"> <u></u><u></u></span></p></div><div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div> 
```

post Sanitization:
```
<div lang="EN-US">Rasmita01<div><div><div><div>Rasmita02<div><div><div><div><div>Rasmita03<div>Rasmita04<div><div><div><div>Rasmita05<div><div><div><div>Rasmita06<div>Rasmita07<div>Rasmita08<div>Rasmita09<div>Rasmita010<div><div><div><div><div><div>Rasmita011<div><div>Rasmita012<div><div><div>Rasmita00<div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div><div>Rasmita000<div>Rasmita0000<div><div><div>Rasmita1 \r\n </div><div></div>Testing Sanitizer is removing divs when the number of open tags are very high<div></div><div><p class="MsoNormal" style="text-align:right;unicode-bidi:embed;direction:rtl" dir="RTL"><span lang="HE"> <u></u><u></u></span></p></div><div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div><div></div><div></div><div></div></div> 
```

See here post Rasmita06 all other divs are not removed.

[Sample_html.txt](https://github.com/OWASP/java-html-sanitizer/files/4855553/Sample_html.txt)
Added the html causing issue to the bug.

**Comment by mikesamuel** [comment]
I think [setNestingLimit](https://www.javadoc.io/static/com.googlecode.owasp-java-html-sanitizer/owasp-java-html-sanitizer/20191001.1/org/owasp/html/TagBalancingHtmlStreamEventReceiver.html#setNestingLimit%28int%29) is what you want.

**Comment by jacobrshields** [comment]
> I think [setNestingLimit](https://www.javadoc.io/static/com.googlecode.owasp-java-html-sanitizer/owasp-java-html-sanitizer/20191001.1/org/owasp/html/TagBalancingHtmlStreamEventReceiver.html#setNestingLimit%28int%29) is what you want.

Correct me if I'm wrong, but there doesn't seem to be a way to change this with the public API.

`setNestingLimit` is called on an instance of `TagBalancingHtmlStreamEventReceiver` that is constructed inside of the implementation of `HtmlSanitizer#sanitize`, so there is no way to access it.

One could try to reimplement their own version of `HtmlSanitizer#sanitize`. However, that method relies on the `HtmlLexer` class, which is not public, so there is no way to access it without reflection.

**Comment by mikesamuel** [comment]
@jacobrshields IIRC, the lexer is invoked by Sanitizer.sanitize which you can supply with a policy that chains together the balancer with other bits and bobs if you're so inclined.

Some of the example code sets up custom pipelines:

https://github.com/OWASP/java-html-sanitizer/blob/fd6b2ddbf7ded2ba4fc1d718e450dc701368cdba/src/main/java/org/owasp/html/examples/EbayPolicyExample.java#L213-L231

**Comment by jacobrshields** [comment]
@mikesamuel I'm not sure I'm following completely.

You can configure your policy with your own `HtmlStreamEventReceiver`, and you can specify your own `HtmlStreamEventProcessor` when you call `HtmlSanitizer#sanitize`, but ultimately a `TagBalancingHtmlStreamEventReceiver` with a nesting limit of `256` is hard-coded to be wrapped around any preprocessor you pass in:

https://github.com/OWASP/java-html-sanitizer/blob/fd6b2ddbf7ded2ba4fc1d718e450dc701368cdba/src/main/java/org/owasp/html/HtmlSanitizer.java#L228-L253

Are you saying you could define an `HtmlStreamEventReceiver` or an `HtmlStreamEventProcessor` such that they hack around the hard-coded `TagBalancingHtmlStreamEventReceiver`?

**Comment by log2akshat** [comment]
In Zimbra we have also faced the same issue, so fixed up by ramping it's value:
https://github.com/Zimbra/java-html-sanitizer-release-20190610.1/commit/bc58f5843b0f5d05d89e94ccd146f21aeb18df81

**Comment by jmanico** [comment]
Thanks for the report, and apologies that it took this long. Your diagnosis was right and the answer you got was not.

#429 fixes the underlying problem, though not the way the thread was heading.

**The nesting limit was never the real issue.** Past 256 elements the sanitizer was *deleting* your content rather than flattening it:

```
300 nested divs around some text
before: 256 open tags, 256 close tags, and the text gone
after:  256 open tags, 256 close tags, and the text between them
```

That is exactly what you described — "sanitizer removes the tags which actually has the content" — and it is why @mikesamuel's `setNestingLimit` pointer did not help, and why @jacobrshields was right that there was no public way to reach it anyway. Raising the limit would only have moved the cliff: you would have lost content at 1025 instead of 257.

The limit is a browser-fidelity number rather than a safety one — its own comment derives 256 from Opera's observed table nesting and WebKit's parser depth cap of 512 — and no HTML parser deletes text when it hits a depth limit. I ran your shape through validator.nu and it keeps both the full depth and the text. So the old behaviour diverged from the parsers the number was chosen to imitate, in the destructive direction.

Your Gmail-quote email now round-trips with the content intact; the wrappers past 256 are flattened, which is what a browser does with them anyway.

**Second, you can now detect it.** A tag dropped for exceeding the limit is reported to `HtmlChangeListener`, so 300 nested divs reports 44 discarded `div` tags. Previously this was completely invisible: the tag balancer runs upstream of the policy, and `HtmlChangeReporter` spots a discarded tag by watching for one that goes into the policy and does not come out — a tag the balancer drops never goes in.

**What I have not done** is make the limit configurable, which is what the title asks for. Now that content is no longer lost, that is a fidelity preference rather than a correctness fix, and it deserves its own decision — including whether to cap how high a caller may raise it. If you still want the knob for your use case, say so and I will open a separate issue for it; I suspect the flattening alone may be enough.

Closing on the data-loss fix. Thanks for the detailed report and the sample HTML — the reproduction shape is now a test.
